manual: Document the use of `static` as a proc call

kaushalmodi opened the pull request:
Also adds tests.

Fixes https://github.com/nim-lang/Nim/issues/16987 .

/cc @hlaaftana PTAL

kaushalmodi committed:
manual: Document the use of `static` as a proc call

Also adds tests.

Fixes https://github.com/nim-lang/Nim/issues/16987 .

konsumlamm commented on doc/manual.rst:
```diff
@@ -2975,6 +2975,21 @@ Even some code that has side effects is permitted in a static block:
   static:
     echo "echo at compile time"
 
+When `static` is used like a procedure call with a proc call as an argument,
+that proc call gets executed at compile-time.
```
```suggestion
`static` can also be used like a routine.
```
Your version implies that the argument has to be a proc call, but it can be any expression.

konsumlamm commented on doc/manual.rst:
```diff
@@ -2975,6 +2975,21 @@ Even some code that has side effects is permitted in a static block:
+
+  proc getNum(a: int): int = a
+
+  # Below calls "echo getNum(123)" at compile time.
+  static:
+    echo getNum(123)
+
+  # Below call evaluates the "getNum(123)" at compile time, but its
+  # result gets used at run time.
+  echo getNum(123).static
```
```suggestion
  # The following expressions are equivalent:

  static:
    getNum(123)

  getNum(123).static

  static(getNum(123))
```
I think it makes more sense to show that using `static` like a routine is equivalent to using a `static:` block.

kaushalmodi replied:
This change will mislead the readers just as it misled me [here](https://github.com/nim-lang/Nim/issues/16987#issuecomment-955853383).

I intentionally did not put all the callable style in that section of the manual as we already have a section to deal with UFCS.

Araq replied:
IMO we shouldn't document that `x.static` works, it's an accidental feature.

kaushalmodi replied:
Do you mean that only the `x.static` style working is accidental, or everything documented in the test in this PR (e.g. `static(x)`?

Can you update the original issue this PR is attempting to close?

Araq replied:
Both `x.static` and `x.static()` are not supposed to work, `static(x)` should be documented.

kaushalmodi replied:
@Araq Done. Please look at the latest collective diff of this PR.

konsumlamm replied:
> Both `x.static` and `x.static()` are not supposed to work, `static(x)` should be documented.

@Araq What's the reasoning behind this? It works for "normal" function call syntax, so why not for MCS as well?

Araq replied:
I should have worded this better: It's not supposed to "not work" either, it's just that I don't know the consequences of allowing it. We already have restrictions on the method call syntax for technical reasons and this one feels like it could be affected too in alternative/future implementations.

kaushalmodi committed:
Update doc/manual.rst

Co-authored-by: konsumlamm <44230978+konsumlamm@users.noreply.github.com>

kaushalmodi committed:
Use the "bug #NNNN" comment syntax for consistency

Ref:
https://nim-lang.github.io/Nim/contributing.html#writing-tests-stdlib

> Always refer to a GitHub issue using the following exact syntax: bug
for tooling.

kaushalmodi committed:
manual: Undocument usage of foo.static

foo.static and foo.static() are not expected to work.

Ref: https://github.com/nim-lang/Nim/pull/19084/files#r741203578
